colcon_python_project/task/ament_python/project/build.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2016-2018 Dirk Thomas
+# Licensed under the Apache License, Version 2.0
+
+# Taken (nearly) verbatim from colcon_ros.task.ament_python.build
+
+import os
+from pathlib import Path
+import sys
+
+from colcon_core.logging import colcon_logger
+from colcon_core.plugin_system import satisfies_version
+from colcon_core.shell import create_environment_hook
+from colcon_core.shell import get_command_environment
+from colcon_core.task import create_file
+from colcon_core.task import install
+from colcon_core.task import TaskExtensionPoint
+from colcon_core.task.python import get_data_files_mapping
+from colcon_core.task.python import get_setup_data
+from colcon_python_project.task.python.project.build \
+    import PythonProjectBuildTask as PythonBuildTask
+
+logger = colcon_logger.getChild(__name__)
+
+
+class AmentPythonProjectBuildTask(TaskExtensionPoint):
+    """Build ROS packages with the build type 'ament_python.project'."""
+
+    def __init__(self):  # noqa: D107
+        super().__init__()
+        satisfies_version(TaskExtensionPoint.EXTENSION_POINT_VERSION, '^1.0')
+
+    async def build(self):  # noqa: D102
+        args = self.context.args
+        logger.info(
+            "Building ROS package in '{args.path}' with build type "
+            "'ament_python.project'".format_map(locals()))
+
+        # reuse Python build task with additional logic
+        extension = PythonBuildTask()
+        extension.set_context(context=self.context)
+
+        # additional hooks
+        additional_hooks = create_environment_hook(
+            'ament_prefix_path', Path(args.install_base),
+            self.context.pkg.name, 'AMENT_PREFIX_PATH', '', mode='prepend')
+
+        # get options from the Python manifest
+        try:
+            env = await get_command_environment(
+                'setup_py', args.build_base, self.context.dependencies)
+        except RuntimeError as e:
+            logger.error(str(e))
+            return 1
+        setup_py_data = get_setup_data(self.context.pkg, env)
+
+        # check if the package index and manifest are being installed
+        data_files = get_data_files_mapping(
+            setup_py_data.get('data_files', []) or [])
```

Comment:
I wonder if we need an alternative approach of getting the package index and manifest, or otherwise not warn about the fallback and it being removed in the future. 

If you migrate away from the legacy `setup.py` build you most likely don't want to have a `setup.py` in your package. My testing showed that this is not possible. From what I understand, this is to keep compatibility with legacy builds or versions of tools that don’t support certain packaging standards (see [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setupcfg-caveats)). If this is the case, I think it'd be great to follow the documentation and to try to keep it a simple `setup()` call.

With this current implementation however the package index and manifest can only be read if you have a setup.py file with its `data_files=[('share/ament_index/resource_index/packages', ['resource/' + package_name]), ('share/' + package_name, ['package.xml'])]` mapping. Otherwise you'll be prompted with the warnings about the fallback and it being removed in the future. I'm unsure what's the best way to solve this and would love to hear your thoughts about this. From what I understand, the traditional data files support is deprecated (see [setuptools documentation](https://setuptools.pypa.io/en/latest/userguide/datafiles.html)).